pkgs/build-support/rust/build-rust-package/default.nix

```
@@ -15,145 +15,149 @@
-
-  cargoBuildNoDefaultFeatures = buildNoDefaultFeatures;
-
-  cargoCheckNoDefaultFeatures = checkNoDefaultFeatures;
-
-  cargoBuildFeatures = buildFeatures;
-
-  cargoCheckFeatures = checkFeatures;
-
-  patchRegistryDeps = ./patch-registry-deps;
-
-  nativeBuildInputs = nativeBuildInputs ++ [
-    cacert
-    git
-    cargoBuildHook
-    cargoCheckHook
-    cargoInstallHook
-    cargoSetupHook
-    rustc
-  ];
-
-  buildInputs = buildInputs
+# depsExtraArgs, cargoLock have to be removed otherwise 'cannot coerce a set to a string'
+# with structuredAttrs this won't be needed
+# also cargoUpdateHook to avoid changing hashes
+(stdenv.mkDerivation (removeAttrs args [ "depsExtraArgs" "cargoLock" "cargoUpdateHook" ])).overrideAttrs (finalAttrs: previousAttrs:
+  let
+    # If we have a cargoSha256 fixed-output derivation, validate it at build time
+    # against the src fixed-output derivation to check consistency.
+    # NOTE: this seems unused
+    #validateCargoDeps = finalAttrs ? cargoHash || finalAttrs ? cargoSha256;
+
+    target = rust.toRustTargetSpec stdenv.hostPlatform;
+    targetIsJSON = lib.hasSuffix ".json" target;
+    useSysroot = targetIsJSON && !__internal_dontAddSysroot;
+
+    # see https://github.com/rust-lang/cargo/blob/964a16a28e234a3d397b2a7031d4ab4a428b1391/src/cargo/core/compiler/compile_kind.rs#L151-L168
+    # the "${}" is needed to transform the path into a /nix/store path before baseNameOf
+    shortTarget =
+      if targetIsJSON then
+        (lib.removeSuffix ".json" (builtins.baseNameOf "${target}"))
+      else target;
+    sysroot = callPackage ./sysroot { } {
+      inherit target shortTarget;
+      RUSTFLAGS = finalAttrs.RUSTFLAGS or "";
+      originalCargoToml = finalAttrs.src + /Cargo.toml; # profile info is later extracted
+    };
+
+    cargoDeps =
+      if cargoVendorDir != null then null
+      else if cargoLock != null then importCargoLock cargoLock
+      else
+        fetchCargoTarball ({
+          src = finalAttrs.src or "";
+          srcs = finalAttrs.srcs or null;
+          sourceRoot = finalAttrs.sourceRoot or null;
+          unpackPhase = finalAttrs.unpackPhase or null;
+          cargoUpdateHook = finalAttrs.cargoUpdateHook or args.cargoUpdateHook or "";
+          # Name for the vendored dependencies tarball
+          name = finalAttrs.cargoDepsName or finalAttrs.name or "${finalAttrs.pname}-${finalAttrs.version}";
+          # see comment below
+          #patches = finalAttrs.cargoPatches;
+          patches = finalAttrs.cargoPatches or [ ];
+        } // lib.optionalAttrs (finalAttrs ? cargoHash) {
+          hash = finalAttrs.cargoHash;
+        } // lib.optionalAttrs (finalAttrs ? cargoSha256) {
+          sha256 = finalAttrs.cargoSha256;
+        } // (args.depsExtraArgs or { }));
+
+    maybeSetStr = x: lib.optionalString (previousAttrs ? ${x}) previousAttrs.${x};
+    maybeSetListFromPrevious = x: lib.optionals (previousAttrs ? ${x}) previousAttrs.${x};
```

is there a specific reason to use these instead of
```suggestion
    maybeSetStr = x: previousAttrs.${x} or "";
    maybeSetListFromPrevious = x: previousAttrs.${x} or [];
```
or inlining them outright? looks like it'd be equivalent and eval slightly faster